### ISSUE TITLE
How to reduce training error 

### ISSUE THREAD

**Issue by theone3nu** [BrainJS/brain.js#654]
BrainJs is really awesome for javascript developers. Thanks for this library. Its working as expected but i have some couple of questions that i am unsure how to fix this.
I am using LSTMTimeStep to predict the next price of currency market.  I have created this 
const net = new brain.recurrent.LSTMTimeStep({
    inputSize: 4,
    hiddenLayers: [32, 32],
    outputSize: 4,
  });
net.train(trainingData, {
    iterations: 50000,
    learningRate: 0.001,
    errorThrash: 0.001,
    logPeriod: 100,
    log: stats => console.log(stats),
  });
I have large data starting from year 2016 daily open high close low values. I have normalised and denormalised as suggested from the tutorial. For every train which i am running on mac 16gb ram it is taking 3 to 4 hours to reach error threshold of 0.021. My question here is there any way to reduce the training error further to 0.01. I have tried changing hidden layers , learning rate, iterations, errorThrash but still my training error is not going below 0.020. 
Also every time when i add new daily data every train it takes 4 to 5 hours to predict next value. I am sure there is a better solution to fix these problems but am unsure how to do it. I saw using multhreading it will be faster but how to do that any example will be helpful.

**Comment by robertleeplummerjr** [comment]
What does your training data look like?

**Comment by theone3nu** [comment]
Hi robert thanks for the reply. Here is my training data looks like. I have chunked the array to 5 elements Array of array just like as you mentioned in the tutorial.
<img width="945" alt="Screen Shot 2020-10-30 at 12 19 28 AM" src="https://user-images.githubusercontent.com/22262757/97619031-bd8afc00-1a45-11eb-91b0-df06e1d3cd45.png">

**Comment by robertleeplummerjr** [comment]
Have you looked into https://github.com/BrainJS/brain.js#cross-validation?

**Comment by theone3nu** [comment]
No i haven't tried this but after seeing this got confused where to use LSTMTimeStep in that cross validation example it is using brain.NeuralNetwork.  Instead of ```
const crossValidate = new brain.CrossValidate(
  brain.NeuralNetwork,
  networkOptions
); 
```can i use` new brain.CrossValidate(brain.recurrent.LSTMTimeStep, networkOptions)`

**Comment by robertleeplummerjr** [comment]
I believe that is the exact usage.

**Comment by theone3nu** [comment]
Ok thanks let me try that. One more question what about 

`const json = crossValidate.toJSON(); // all stats in json as well as neural networks`
```
const net = crossValidate.toNeuralNetwork(); // get top performing net out of `crossValidate`

// optionally later
const json = crossValidate.toJSON();
const net = crossValidate.fromJSON(json);

```
Where should i use the above one

**Comment by theone3nu** [comment]
I just tried using crossValidate.   Here is my code 
```
`const networkOptions = {
    inputSize: 4,
    hiddenLayers: [32, 32],
    outputSize: 4,
  };
  const trainingOptions = {
    iterations: 20000,
    learningRate: 0.001,
    errorThrash: 0.018,
    logPeriod: 100,
    log: stats => console.log(stats),
  };
`
`const crossValidate = new brain.CrossValidate(
    brain.recurrent.LSTMTimeStep,
    networkOptions
  );
  const stats = crossValidate.train(trainingData, trainingOptions);
  console.log(stats);
  const net = crossValidate.toNeuralNetwork();`
```

I see there is much more error using crossValidate than without using crossValidate. Also Time taken is high same like without crossValidate. Let me know anything wrong in the above implementation?

**Comment by theone3nu** [comment]
Any update on this please. I tried using crossValidate but still no change in terms of time and error rate. Any suggestions please?

**Comment by robertleeplummerjr** [comment]
Maybe you could normalize values with a cap much lower than you are currently?  It seems none of the values in training are above 0.2.  If they were a greater the net can more easily pick up on the values being trained on.

**Comment by theone3nu** [comment]
Currently this is the formula i am using for scaling down and scaling up.  Any other formula to normalize values. The normalized data range should be from 0 to 1 right thats what that i achieved in trainingdata.
```

`  function scaleDown(step) {
    return {
      open: (step.open - lowest) / (highest - lowest),
      high: (step.high - lowest) / (highest - lowest),
      low: (step.low - lowest) / (highest - lowest),
      close: (step.close - lowest) / (highest - lowest),
    };
  }
  function scaleUP(step) {
    return {
      open: step.open * (highest - lowest) + lowest,
      high: step.high * (highest - lowest) + lowest,
      low: step.low * (highest - lowest) + lowest,
      close: step.close * (highest - lowest) + lowest,
    };
  }
`
```

**Comment by robertleeplummerjr** [comment]
How many values are above the highest 33% for example?  You may get away with just ignoring these spikes.

**Comment by theone3nu** [comment]
Not sure how to calculate  how many values that  are above the highest 33%. Any example do you have.

**Comment by robertleeplummerjr** [comment]
https://observablehq.com/@robertleeplummerjr/how-many-values-are-above-33

**Comment by theone3nu** [comment]
Thanks for the response. There are 0 values which are above 33% as all my sacaled data values are between 0 and 1.  See the values below.

<img width="928" alt="Screen Shot 2020-10-30 at 10 41 09 PM" src="https://user-images.githubusercontent.com/22262757/97735987-0f448c80-1b01-11eb-967b-1ca1669b043c.png">

**Comment by robertleeplummerjr** [comment]
I meant denormalized values, not normalized ones.

**Comment by theone3nu** [comment]
Denormalized values are original values which are in betweeen 101.17932 and 118.665 so all values are above 33%. 
<img width="768" alt="Screen Shot 2020-10-30 at 10 58 23 PM" src="https://user-images.githubusercontent.com/22262757/97738046-d78b1400-1b03-11eb-81e6-79a3e526b816.png">

**Comment by robertleeplummerjr** [comment]
What values have you used for learning rate?

**Comment by robertleeplummerjr** [comment]
Also, have you used more, perhaps smaller, layers?

**Comment by theone3nu** [comment]
This is my training options data 

` net.train(trainingData, {
    iterations: 20000,
    learningRate: 0.001,
    errorThrash: 0.0001,
    logPeriod: 100,
    log: stats => console.log(stats),
  });`
and 

` const net = new brain.recurrent.LSTMTimeStep({
    inputSize: 4,
    hiddenLayers: [32, 32],
    outputSize: 4,
  });`
i have changed to smaller layers but errror rate is not reaching 0.02 so i just tried to increase layers. Even though i tried with small layers it is taking same time and error rate not much difference. Perhaps multi threading will fix this issue. But LSTM Time stamp multithreading is not there. Any suggestions please. For each train time taking is 4 to 5 hours almost.

**Comment by theone3nu** [comment]
I have fixed my two problems by storing the training net and reusing it again. Calling repeatedly to reduce error rate to much better from 0.020 to 0.015. Also when i add new data i am just appending it to the trained net it is continuing from there with faster results. 
```
`const net = new brain.recurrent.LSTMTimeStep({
    inputSize: 4,
    hiddenLayers: [16, 16],
    outputSize: 4,
  });

  net.train(trainingData, {
    iterations: 10000,
    learningRate: 0.001,
    errorThrash: 0.001,
    logPeriod: 100,
    log: stats => console.log(stats),
  });
  const json = net.toJSON()
// write to file system
  fs.writeFileSync('trained-net.json', JSON.stringify(json));`
```


```
`
//Later with new data loading the trained net and adding new data with less iterations.
const json2= JSON.parse(fs.readFileSync('eurusddailytrained-net.json'))
  const trainedNet = new brain.recurrent.LSTMTimeStep()
  trainedNet.fromJSON(json2);
trainedNet.train(trainingData, {
    iterations: 2000,
    learningRate: 0.001,
    errorThrash: 0.001,
    logPeriod: 100,
    log: stats => console.log(stats),
  });
  const json = trainedNet.toJSON()
// write to file system
  fs.writeFileSync('trained-net.json', JSON.stringify(json));
`
```
If i keep on re run error rate is reducing and more accurate results are coming.  BrainJs is awesome. Thanks robert.

**Comment by robertleeplummerjr** [comment]
This is fantastic!

**Comment by davidmigithub** [comment]
This thread is old and apparently solved by the opener, but I'm seeing and error in the training options, you put errorThrash: instead of errorThresh:

**Comment by Stephen-Montague** [comment]
This thread is interesting as a reference point for many Brain.js topics, but on the topic of speeding up training - I noticed in the thread above there were logging calls every 100 iterations, and I can't say for sure until I do some tests here, but in general, logging out status to a file or terminal often can really hit speed.  Would just log however much you absolutely need, to gauge progress, otherwise, have patience and wait for the final result.
